src/mem/Allocator.h

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (C) 1996-2022 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID_SRC_MEM_ALLOCATOR_H
+#define SQUID_SRC_MEM_ALLOCATOR_H
+
+#include "mem/forward.h"
+
+namespace Mem
+{
+
+/**
+ * Basic API definition for memory pooling allocators.
+ *
+ * A pool is a [growing] space for objects of the same size
+ */
```

Comment:
IMO the pooling mention should stay. The distinction of this API from, eg the STL allocators, is that it explicitly plugs the configurable Squid "Memory Pools" into the allocator logic. Without the pooling ability (even when disabled/unused) there would be no point in Squid providing its own allocator(s).

The part that I think we should discard is the "same size" / "fixed-size" wording from the old definition. Since the child PoolMalloc already deals with non-same size objects and the "Memory Pools" documentation should be where the definition of a pool is.

Reply:
> The distinction of this API from, eg the STL allocators, is that it explicitly plugs the configurable Squid "Memory Pools" into the allocator logic.

I see nothing about pools in this API. There are a couple of words "pool" sprinkled around for historical reasons, but the things they name represent pooling-agnostic concepts (or what should be pooling-agnostic concepts).

> Without the pooling ability there would be no point in Squid providing its own allocator(s).

I disagree. The statistics maintained by Allocators is very valuable without any pooling. And there is also memory zero-ing functionality.


> The part that I think we should discard is the "same size" / "fixed-size" wording from the old definition. Since the child PoolMalloc already deals with non-same size objects

MemPoolMalloc does not "deal with non-same size objects" AFAICT. It allocates `obj_size`-byte objects. Am I missing something?

> and the "Memory Pools" documentation should be where the definition of a pool is.

Agreed, but "fixed-size objects" is not about pooling.

Reply:
The hierarchy of classes for which this is the base/root class _is_ "memory pools". Each class of this hierarchy is providing various ways to access a pool instance. To be clear what I am proposing is to keep the line I added in this PR
, with maybe some slight modifications like so:
 * "Basic API definition for memory pooling allocators."
 * "Basic API definition for Memory Pools allocators."
 
What you are missing about PoolMalloc is that the obj_size it allocates can be updated. 




> > The distinction of this API from, eg the STL allocators, is that it explicitly plugs the configurable Squid "Memory Pools" into the allocator logic.
> 
> I see nothing about pools in this API. There are a couple of words "pool" sprinkled around for historical reasons, but the things they name represent pooling-agnostic concepts (or what should be pooling-agnostic concepts).
> 

Exactly why mention of memory pools should stay in the description. To clarify that this API is how pool allocators are accessed by code not caring about how the pool operates.


> > Without the pooling ability there would be no point in Squid providing its own allocator(s).
> 
> I disagree. The statistics maintained by Allocators is very valuable without any pooling. And there is also memory zero-ing functionality.
> 

In general that might be true, but the statistics implemented by Squid are all about pools. Memory zero'ing logic is also all about pools. Non-pooling would (and does) just use calloc().

> > The part that I think we should discard is the "same size" / "fixed-size" wording from the old definition. Since the child PoolMalloc already deals with non-same size objects
> 
> MemPoolMalloc does not "deal with non-same size objects" AFAICT. It allocates `obj_size`-byte objects. Am I missing something?
> 

Sorry, I was crossing PoolMalloc with the xmalloc pools. To remove the old C-style buffer/strings/rigid we will need a pool class that accepts multiple sizes of object and manages them. I am thinking that would be PoolMalloc growing ability to provide re-alloc.

Reply:
> The hierarchy of classes for which this is the base/root class _is_ "memory pools".

I disagree. It is a hierarchy of memory allocators. Pooling is just one of the optional features some allocators are capable of.

> Each class of this hierarchy is providing various ways to access a pool instance. 

Each leaf class implements a memory allocator interface. Implementation details and extras that vary from class to class are not important when we define that interface.

> To be clear what I am proposing is to keep the line I added in this PR , with maybe some slight modifications like so:
> * "Basic API definition for memory pooling allocators."
> * "Basic API definition for Memory Pools allocators."

We should document the interface declared by the Allocator class. That interface has nothing to do with pooling. That interface clients/users do not care about pooling. Even if every single leaf class implements some form of pooling, it would be still wrong to document this base interface as a pooling allocator interface. There is no reason to add things to the interface description that the interface itself and its users/clients do not care about, especially when those things may be disabled.


> > > The distinction of this API from, eg the STL allocators, is that it explicitly plugs the configurable Squid "Memory Pools" into the allocator logic.

> > I see nothing about pools in this API. There are a couple of words "pool" sprinkled around for historical reasons, but the things they name represent pooling-agnostic concepts (or what should be pooling-agnostic concepts).

> Exactly why mention of memory pools should stay in the description. To clarify that this API is how pool allocators are accessed by code not caring about how the pool operates.

The suggested wordings are not _clarifications_ IMO. They mislead the reader and may be in the way of adding special allocators that do not pool at all. However, I support mentioning pooling in the following (or similar) non-interface-describing way. It may help with code navigation and such.

```suggestion
/// An interface for memory allocators that deal with fixed-size objects.
/// Allocators may optimize repeated de/allocations using memory pools.
```

> > > Without the pooling ability there would be no point in Squid providing its own allocator(s).
> > 
> > 
> > I disagree. The statistics maintained by Allocators is very valuable without any pooling. And there is also memory zero-ing functionality.
> 
> In general that might be true, but the statistics implemented by Squid are all about pools.

The statistics implemented by Squid has very valuable parts that are not related to pooling in any way. One may disable pooling and still get valuable memory allocation stats.

> Memory zero'ing logic is also all about pools.

Memory zeroing is about preparation of allocated (or freed) object memory. It is not related to pooling in any way.

> Non-pooling would (and does) just use calloc().

The existence of multiple ways to allocate, zero, and pool object memory does not imply that there is no point in Squid providing its own allocators without pooling.

> To remove the old C-style buffer/strings/rigid we will need a pool class that accepts multiple sizes of object and manages them.

I am not sure that is the right path forward, but we do not need to agree on that right now. If it turns out to be the right path, we can adjust the Allocator API accordingly, of course.

Reply:
Taking your updated suggestion as "good enough". The remainder of the points we disagree on are perhapse best resolved later in PR relevant to those points of discussion. That last point in particular I am moving to the PR about AllocatorProxy.